### PR TITLE
Prevent NMO from being deployed on master nodes

### PR DESCRIPTION
**Pull request by omeryahud**
This PR proposes to prevent NMO pods from being deployed on master nodes, as it is a bad practice to deploy 3rd party apps on master nodes in OCP.

### CODE REVIEW

**Comment by yanirq** [comment]
This calls for a test plane and a CI e2e test to see what happens when a node with NMO is drained

**Comment by dankenigsberg** [comment]
@omeryahud why do you need this PR? Please include your motivation in every commit message and every description of every PR.

**Comment by beekhof** [comment]
We need to know the context for this change before we can consider it.
Doesn't make much sense at first glance.

**Comment by omeryahud** [comment]
@yanirq Yes you are correct, I just opened this PR to see if any of the test lanes would fail.
@dankenigsberg @beekhof The motivation for this PR is to prevent NMO from being deployed on master nodes, because it is considered a bad practice to deploy 3rd party apps on master nodes in OCP.
I opened this PR just to see if any of the test lanes would fail, I suppose I should have declared it as Work in Progress.

**Comment by beekhof** [comment]
I would argue that it shouldn't be a 3rd party operator, but that's a different conversation.

IMHO the complexity of the NMO potentially evacuating itself trumps the "no masters" best practice in this case.  It would make the admin experience demonstrably worse.

**Comment by omeryahud** [comment]
If I'm not wrong, the e2e tests in NMO test exactly that, that NMO manages to drain the node it's deployed on. 
In any case, I would argue that the place for NMO in OCP is on an infra node.
With regards to k8s, in order to drain a node, you must have at least one other node that is schedulable, so in case NMO is deployed on the drained node, it would just redeploy on the other node and continue draining the target node, this is the process I am seeing with manual testing atm.

**Comment by omeryahud** [comment]
@yanirq @beekhof @ksimon1 could any you guys take a look please?

File: test/e2e/nodemaintenance_test.go [hunk 1]
```diff
@@ -354,3 +362,74 @@ func checkFailureStatus(t *testing.T, f *framework.Framework) {
...
+	for _, pod := range pods.Items {
+		if !sliceContainsString(hostnames, pod.Spec.NodeName) {
+			hostnames = append(hostnames, pod.Spec.NodeName)
+		}
+	}
+
+	return hostnames
+}
+
+func sliceContainsString(slice []string, str string) bool {
```
Comment (yanirq) [inline test/e2e/nodemaintenance_test.go]: maybe create a utils folder/class for that under test folder and place this function there , as in:
https://github.com/kubevirt/node-maintenance-operator/blob/master/pkg/controller/nodemaintenance/utils.go

File: test/e2e/nodemaintenance_test.go [hunk 2]
```diff
@@ -354,3 +362,74 @@ func checkFailureStatus(t *testing.T, f *framework.Framework) {
...
+
+	if len(pods.Items) < 1 {
+		t.Fatal("there are no operator pods")
+	}
+
+	return pods.Items[0].Spec.NodeName
+}
+
+func countWorkerNodes(t *testing.T, f *framework.Framework) int {
+	nodes := corev1.NodeList{}
```
Comment (yanirq) [inline test/e2e/nodemaintenance_test.go]: another suggestion to get all nodes with a label:
```
// GetByRole returns all nodes with the specified role
func GetByRole(c client.Client, role string) ([]corev1.Node, error) {
	selector, err := labels.Parse(fmt.Sprintf("%s/%s=", testutils.LabelRole, role))
	if err != nil {
		return nil, err
	}

	nodes := &corev1.NodeList{}
	if err := c.List(context.TODO(), nodes, &client.ListOptions{LabelSelector: selector}); err != nil {
		return nil, err
	}
	return nodes.Items, nil
}
```

Can move to a util class

Comment (omeryahud) [reply test/e2e/nodemaintenance_test.go]: Problem is, I don't think worker nodes have roles applied (or not all of the time):
```
$ kubectl get nodes
NAME     STATUS   ROLES    AGE   VERSION
node01   Ready    master   20h   v1.16.2
node02   Ready    <none>   20h   v1.16.2
node03   Ready    <none>   20h   v1.16.2
```

File: test/e2e/nodemaintenance_test.go (hunk 1 above)
Comment (omeryahud) [reply test/e2e/nodemaintenance_test.go]: Yea, sounds good, will do

File: test/e2e/nodemaintenance_test.go (hunk 2 above)
Comment (yanirq) [reply test/e2e/nodemaintenance_test.go]: right.  well if some of the code lines here are more elegant feel free to use them anyway

**Comment by omeryahud** [comment]
@beekhof @ksimon1 could you guys take a look please?

**Comment by beekhof** [comment]
I don't doubt that it's a good implementation, but I'm not seeing that there is a compelling reason for this change.

Is there anything you can point that illustrates this operator running on a master is causing a problem?  

IMHO, it is a deficiency that the NMO is not part of OCP and that masters are exactly where the NMO _should_ be running.

**Comment by omeryahud** [comment]
> Is there anything you can point that illustrates this operator running on a master is causing a problem?

Nothing other than OCP best practices.

> IMHO, it is a deficiency that the NMO is not part of OCP and that masters are exactly where the NMO _should_ be running.

Who should be involved and come up with the decision to make it part of OCP?

**Comment by fabiand** [comment]
I suppose the easy answer is: NMO is not a master.
Technically, and please keep me honest, it should not matter if NMO is running on a worker or master.

A servicability reason for not running NMO on masters is simply to make and keep the resource consumption of masters predictable by leaving those nodes to core master components.

Certain K8s distributions like OCP also have concepts like infra nodes, which are hosting operators, that is actually where I'd imagine NMO to live.
Or ideally - on the long run - become a part of the core of Kubernetes ... let the dreams live on.

**Comment by fabiand** [comment]
ping?

**Comment by slintes** [comment]
Hey @fabiand, small world :wink: 

I would be happy to drive this PR forward, but from the comments I'm not sure if there is an agreement on how to go on.
My take: 
> A servicability reason for not running NMO on masters is simply to make and keep the resource consumption of masters predictable by leaving those nodes to core master components.

Because of this I would also say that NMO should not run on masters.
But we should make sure that NMO doesn't take itself down by adding a `PodDisruptionBudget` to its deployment. Does that make sense? @beekhof wdyt?

**Comment by beekhof** [comment]
What does the patch do if there are only masters?
Is a PDB overkill for a single pod?

**Comment by beekhof** [comment]
What I would like to avoid is the UX nightmare that the console has when it runs out of workers to run on.

If we can make it prefer workers, but run on masters when there is no other choice (either there were never workers in the cluster, or they failed), that is something I could get onboard with.

**Comment by fabiand** [comment]
Hello (waving from my continent to the next).

I understand (I think I do) @beekhof concerns, however, if there are no workers, then we will have many issues with other worklaods and operators as well.
I'd recommend to take this to some OpenShift discussion place to hear their opinions.

because technically we can run it anywhere, I'm just looking if we can make the cluster behavior over all more predictable.

**Comment by beekhof** [comment]
Looks like David has a thread called "how to influence operator pod location when installing via olm subscription?", but it doesn't seem like there are good answers.

**Comment by beekhof** [comment]
> What does the patch do if there are only masters?
> Is a PDB overkill for a single pod?

@omeryahud  Do we have an answer for this?

**Comment by slintes** [comment]
The discussion about being installed on master node revived, and we decided to follow kubevirt and friends by removing the affinity, because it also conflicts with a potential node selector on the OLM subscription. Handling this in #140 instead of trying to revive this PR though, sorry.

/close

**Pull request closed without merging**